File: service/lua/sliding_window_log.lua
-- KEYS[1] = zset key
-- ARGV[1] = now (unix seconds, float)
-- ARGV[2] = window_seconds
-- ARGV[3] = limit
-- ARGV[4] = unique member id for this request

local key = KEYS[1]
local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local limit = tonumber(ARGV[3])
local member = ARGV[4]

redis.call('ZREMRANGEBYSCORE', key, 0, now - window)
local count = redis.call('ZCARD', key)

local allowed = 0
if count < limit then
    redis.call('ZADD', key, now, member)
    allowed = 1
    count = count + 1
end
redis.call('EXPIRE', key, window + 1)

return {allowed, count}

File: service/algorithms/sliding_window_log.py
"""
Sliding window log. Every allowed request's timestamp is stored in a
Redis sorted set; on each check, entries older than the window are
trimmed and the remaining count is compared to the limit. Exact - no
approximation error - at the cost of one sorted-set entry per request
for the length of the window.
"""

import uuid
from pathlib import Path
from time import time

from redis.asyncio import Redis

from ..redis_client import eval_script

_SCRIPT = (Path(__file__).resolve().parent.parent / "lua" / "sliding_window_log.lua").read_text()


async def check(client: Redis, key: str, limit: int, window_seconds: float) -> dict:
    now = time()
    member = f"{now}-{uuid.uuid4().hex[:8]}"
    allowed, count = await eval_script(
        client, _SCRIPT, 1, f"swl:{key}", now, window_seconds, limit, member
    )
    return {"allowed": bool(int(allowed)), "count": int(count)}
